### ISSUE TITLE
Failing to compare preprocessing pipelines (UX issue, probably)

### ISSUE THREAD

**Issue by GaelVaroquaux** [probabl-ai/skore#2509]
### What would you like to say?

To demo skore, I tried adapting the gist of a [skrub example](https://skrub-data.org/stable/auto_examples/0020_text_with_string_encoders.html) that compares various preprocessing choices (I'm putting the code at the bottom of the issue).

I'm not finding it easy to get what I want.

1. First, Skore seems not well adapted because (if I understand correctly) in its displays the various estimators summarizing their name to the last supervised estimator of a pipeline:

<img width="1522" height="281" alt="Image" src="https://github.com/user-attachments/assets/d33a4695-86d9-45d6-880d-3a849f901829" />
For the following estimator:

```python
from skrub import StringEncoder
string_encoder = StringEncoder(ngram_range=(3, 4), analyzer="char_wb", random_state=0)

string_encoder_pipe = make_pipeline(
    TableVectorizer(high_cardinality=string_encoder),
    HistGradientBoostingClassifier(),
)
```

2. Second, at the end, I try to use the summary display to compare my preprocessing  

<img width="1845" height="1347" alt="Image" src="https://github.com/user-attachments/assets/f7198a0a-ca40-43a5-85db-835ea45c62ee" />

But this doesn't seem to separate out my preprocessing pipeline

## Code to reproduce:

```python
# %%
# The data
from skrub.datasets import fetch_toxicity
dataset = fetch_toxicity()
X, y = dataset.X, dataset.y
X["is_toxic"] = y

y = X.pop("is_toxic").map({"Toxic": 1, "Not Toxic": 0})

# %%
from sklearn.ensemble import HistGradientBoostingClassifier
from sklearn.pipeline import make_pipeline
from skrub import TableVectorizer



from skore import CrossValidationReport, Project
my_project = Project("my_project")

# %%
# MinHashEncoder
from skrub import MinHashEncoder

minhash_pipe = make_pipeline(
    TableVectorizer(high_cardinality=MinHashEncoder(n_components=30)),
    HistGradientBoostingClassifier(),
)
minhash_results = CrossValidationReport(minhash_pipe, X, y, splitter=3)
my_project.put("MinHashEncoder", minhash_results)

# %%
# StringEncoder
from skrub import StringEncoder
string_encoder = StringEncoder(ngram_range=(3, 4), analyzer="char_wb", random_state=0)

string_encoder_pipe = make_pipeline(
    TableVectorizer(high_cardinality=string_encoder),
    HistGradientBoostingClassifier(),
)

string_encoder_results = CrossValidationReport(string_encoder_pipe,
        X, y, splitter=3)
my_project.put("StringEncoder", string_encoder_results)

# %%

summary = my_project.summarize()
summary
```

**Comment by glemaitre** [comment]
Yep we failed at attempting solving this one in https://github.com/probabl-ai/skore/pull/1771

Just pasting some discussion that we had at the time:

<details>

What do we do so far? To split the pipeline, we output a dict where the keys are the modules, and the values are list of transformers used in the pipeline (ordered alphabetically).  

This solution is not great because the modules may be private (as they are in skrub), and some modules only contain one transformer, therefore they are not interesting. 

Furthermore, it means that for some coordinates, such as `preprocessing`, where many elements may be used in a single pipeline (for instance, several scalers). Therefore, for this kind of coordinates, the available points will be list, potentially quite large. It will renders something that is not pretty at all, and affect usability.

One option could be to have a dict where we map the transformers into a category we find explicit. For instance, the transformers in the module `preprocessing` of scikit-learn would be split into scalers and encoders.

This solution is not satisfactory either because it means that we should maintain this dict, and also because it doesn’t fix the large height and width.

**Conclusion:** So far, we don’t know how to present it nicely, in an exploitable way, in lib. In hub, we could find a solution with filtering as it’s done in e-commerce.

**Decision:** For now, we will drop this feature from the priorities. It doesn’t mean that we will not come back to it later.

</details>

So I personally did not put much more thoughts on the subject apart of thinking about on for the hub. There I thought that we could come with projecting the signature of the pipeline in a text embedding that could tell us how it is.

Happy to discuss and revive this issue.

Here, it could allow use to differentiate the pipeline both the projection would be so hard to make sense out of it.

**Comment by GaelVaroquaux** [comment]
I cannot speak for the technical details, but it seems to me that here we are failing (in a very implicit way) on a perfectly valid usecase, that might not be unfrequent.

I do agree that there is a bit of work in getting a UX "key" of a pipeline that allows to separate and display things in meaningful way. But this work seem important, and I don't see a way out of it

**Comment by GaelVaroquaux** [comment]
In terms of technical solution (my brain has been running, though I said that I would not look at it), here are a few ideas:

- In terms of desired output, heavy use of ellipisis. For instance, when displaying multiple estimators (as in the summary) the case in the above example could lead to "...StringEncoder...", "...MinHashEncoder...". If hovering on the names gives a full repr, it would be great, but I don't know how feasible this is in plotly. When only one estimator is displayed, I would go for its repr.
  - How to get there, I'm not sure, but it could be taking the repr of the estimators, chunking them into words and looking at which differ across the various entries to display. Things like the "SequenceMatcher" in the standardlib can maybe be helpful here

**Comment by glemaitre** [comment]
Thanks @GaelVaroquaux I'll give another round of thoughts with those in mind.

**Comment by auguste-probabl** [comment]
Just to resolve the "every estimator is collapsed to one" problem, maybe we could include the datetime of the `put` as the first axis of the plot, or some other unique identifier (I think we should make the key a skore-defined hash, and let users tag their artifacts only if they want to)

**Comment by GaelVaroquaux** [comment]
> Just to resolve the "every estimator is collapsed to one" problem, maybe we could include the datetime of the put as the first axis of the plot,

The datetime would be very useful and it should be there ( https://github.com/probabl-ai/skore/issues/1997 ), but it is a pretty poor user experience, and the core of the issue should be addressed. The intent of the user is to compare pipeline. The UX should enable it

**Comment by auguste-probabl** [comment]
May be solved/made irrelevant by #2960

**Comment by jeromedockes** [comment]
I don't think so: how to make a very compact description of pipelines that allows to quickly see how they differ is still very much open.

In general it will be difficult, and a user can fall back to looking at the date & time when they were fitted, manually added tags, or other metadata (or even scores) to recognize their pipelines. However when it is possible, it would be useful to produce an informative summary / contrast of the pipelines' actual contents. It may turn out that it is very hard in realistic cases, but still worth putting some effort into.
